File: problem-solving/baekjoon/18258_큐_2.go
// https://www.acmicpc.net/problem/18258
// 큐의 개념을 익히고 실습하는 문제. 연산 당 시간 복잡도가 O(1)이어야 한다는 점에 유의.
package main

import (
	"bufio"
	"fmt"
	"os"
)

func main() {
	reader := bufio.NewReader(os.Stdin)
	writer := bufio.NewWriter(os.Stdout)
	defer writer.Flush()

	var n int
	fmt.Fscanln(reader, &n)

	var queue []int
	for i := 0; i < n; i++ {
		var command string
		var num int
		fmt.Fscanln(reader, &command, &num)

		switch command {
		case "push":
			queue = append(queue, num)
		case "pop":
			if len(queue) > 1 {
				fmt.Fprintln(writer, queue[0])
				queue = queue[1:]
			} else if len(queue) == 1 {
				fmt.Fprintln(writer, queue[0])
				queue = []int{}
			} else {
				fmt.Fprintln(writer, -1)
			}
		case "size":
			fmt.Fprintln(writer, len(queue))
		case "empty":
			if len(queue) == 0 {
				fmt.Fprintln(writer, 1)
			} else {
				fmt.Fprintln(writer, 0)
			}
		case "front":
			if len(queue) > 0 {
				fmt.Fprintln(writer, queue[0])
			} else {
				fmt.Fprintln(writer, -1)
			}
		case "back":
			if len(queue) > 0 {
				fmt.Fprintln(writer, queue[len(queue)-1])
			} else {
				fmt.Fprintln(writer, -1)
			}
		}
	}
}

File: problem-solving/baekjoon/18258_큐_2.py
# https://www.acmicpc.net/problem/18258
# 큐의 개념을 익히고 실습하는 문제. 연산 당 시간 복잡도가 O(1)이어야 한다는 점에 유의.
"""
deque 활용.
Deques support thread-safe, memory efficient appends and pops from either side of the deque 
with approximately the same O(1) performance in either direction.
"""
import sys
from collections import deque   

if __name__ == "__main__":
    n = int(sys.stdin.readline())
    dq = deque()
    for i in range(n):
        inputs = sys.stdin.readline().split()
        command = inputs[0]
        num = 0
        if len(inputs) > 1:
            num = inputs[1]
        if command == "push":
            dq.append(num)
        elif command == "pop":
            if len(dq) > 0:
                print(dq.popleft())
            else:
                print(-1)
        elif command == "size":
            print(len(dq))
        elif command == "empty":
            if len(dq) == 0:
                print(1)
            else:
                print(0)
        elif command == "front":
            if len(dq) > 0:
                print(dq[0])
            else:
                print(-1)
        elif command == "back":
            if len(dq) > 0:
                print(dq[-1])
            else:
                print(-1)

"""
# 시간초과 발생한 코드
# (list slicing 시간복잡도=O(n+k))
import sys

if __name__ == "__main__":
    n = int(sys.stdin.readline())
    queue = []
    for i in range(n):
        inputs = sys.stdin.readline().split()
        command = inputs[0]
        num = 0
        if len(inputs) > 1:
            num = inputs[1]
        if command == "push":
            queue.append(num)   # 시간복잡도: O(1)
        elif command == "pop":
            if len(queue) > 1:
                print(queue[0])
                queue = queue[1:]   # 시간복잡도: O(n+k)
            elif len(queue) == 1:
                print(queue[0])
                queue = []
            else:
                print(-1)
        elif command == "size":
            print(len(queue))
        elif command == "empty":
            if len(queue) == 0:
                print(1)
            else:
                print(0)
        elif command == "front":
            if len(queue) > 0:
                print(queue[0])
            else:
                print(-1)
        elif command == "back":
            if len(queue) > 0:
                print(queue[len(queue)-1])
            else:
                print(-1)
"""